### ISSUE TITLE
angular: unable to use ng2-charts-schematics and @ngrx/* schematics with workspace.json v2

### ISSUE THREAD

**Issue by LayZeeDK** [nrwl/nx#4365]
I wasn't sure whether to file an issue with ng2-charts or Nx, but Zack De Rose and Victor Savkin asked me to file an issue with reproduction steps to look into the matter, so here we go.

## Current Behavior

When using `workspace.json` version 2, I'm unable to use `ng2-charts-schematics` to generate chart components.

If I switch to `angular.json` version 1, I'm able to use `ng2-charts-schematics` to generate chart components.

## Expected Behavior

When using `workspace.json` version 2, I should be able to use `ng2-charts-schematics` to generate chart components.

### Create and configure empty workspace
1. Run `yarn global add @nrwl/cli@11.0.15`
1. Run `yarn global add create-nx-workspace@11.0.15`
1. Run `yarn create nx-workspace nrwl-airlines --npm-scope=nrwl-airlines --preset=empty --linter=eslint --no-nx-cloud --style=css --package-manager=yarn`
1. Run `npx json -I -f workspace.json -e "this.cli.packageManager = 'yarn';"`
1. Run `yarn remove tslint`

### Add Angular capabilities
1. Run `yarn add --dev @nrwl/angular`
1. Run `nx generate @nrwl/angular:init`
1. Run `npx json -I -f workspace.json -e "this.generators['@nrwl/angular:application'].linter = 'eslint';"`
1. Run `npx json -I -f workspace.json -e "this.generators['@nrwl/angular:library'].linter = 'eslint';"`

### Generate Angular projects
1. Run `nx generate @nrwl/angular:application --name=booking-app --prefix=booking --tags="project:booking-app,scope:booking,type:app" --no-interactive`
1. Run `nx generate @nrwl/angular:library --name=feature-flight-search --directory=booking --prefix=booking --tags="project:booking-app,scope:booking,type:feature" --buildable --no-interactive`
1. Run `yarn remove codelyzer`

### Generate chart component
1. Run `yarn add ng2-charts ng2-charts-schematics`
1. Run `nx generate module my-line-chart --project=booking-feature-flight-search`
1. Run `nx generate ng2-charts-schematics:line my-line-chart --project=booking-feature-flight-search --module=my-line-chart`
1. Exception is thrown.

If I change `workspace.json` version 2 into `angular.json` version 1, the `ng2-charts-schematics:line` schematic works.

### Failure Logs

When running `nx generate ng2-charts-schematics:line my-line-chart`, this error is output:
```
SchematicsException [Error]: Could not find (undefined)
    at Object.getWorkspace (C:\Users\larsb\projects\sandbox\nrwl-airlines\node_modules\@schematics\angular\utility\config.js:22:15)
    at Object.getProject (C:\Users\larsb\projects\sandbox\nrwl-airlines\node_modules\@schematics\angular\utility\project.js:20:20)
    at C:\Users\larsb\projects\sandbox\nrwl-airlines\node_modules\@schematics\angular\component\index.js:92:35
    at MergeMapSubscriber.project (C:\Users\larsb\projects\sandbox\nrwl-airlines\node_modules\@angular-devkit\schematics\src\rules\call.js:75:24)
    at MergeMapSubscriber._tryNext (C:\Users\larsb\projects\sandbox\nrwl-airlines\node_modules\@angular-devkit\schematics\node_modules\rxjs\internal\operators\mergeMap.js:67:27)
    at MergeMapSubscriber._next (C:\Users\larsb\projects\sandbox\nrwl-airlines\node_modules\@angular-devkit\schematics\node_modules\rxjs\internal\operators\mergeMap.js:57:18)
    at MergeMapSubscriber.Subscriber.next (C:\Users\larsb\projects\sandbox\nrwl-airlines\node_modules\@angular-devkit\schematics\node_modules\rxjs\internal\Subscriber.js:66:18)
    at Observable._subscribe (C:\Users\larsb\projects\sandbox\nrwl-airlines\node_modules\@angular-devkit\schematics\node_modules\rxjs\internal\util\subscribeToArray.js:5:20)
    at Observable._trySubscribe (C:\Users\larsb\projects\sandbox\nrwl-airlines\node_modules\@angular-devkit\schematics\node_modules\rxjs\internal\Observable.js:44:25)
    at Observable.subscribe (C:\Users\larsb\projects\sandbox\nrwl-airlines\node_modules\@angular-devkit\schematics\node_modules\rxjs\internal\Observable.js:30:22)
Could not find (undefined)
```

### Environment

`nx report` output
```
  Node : 12.20.0
  OS   : win32 x64
  yarn : 1.22.10

  nx : Not Found
  @nrwl/angular : 11.0.15
  @nrwl/cli : 11.0.15
  @nrwl/cypress : 11.0.15
  @nrwl/devkit : 11.0.15
  @nrwl/eslint-plugin-nx : 11.0.15
  @nrwl/express : Not Found
  @nrwl/jest : 11.0.15
  @nrwl/linter : 11.0.15
  @nrwl/nest : Not Found
  @nrwl/next : Not Found
  @nrwl/node : Not Found
  @nrwl/react : Not Found
  @nrwl/schematics : Not Found
  @nrwl/tao : 11.0.15
  @nrwl/web : Not Found
  @nrwl/workspace : 11.0.15
  typescript : 4.0.5
```

`workspace.json` version 2 (doens't work with this schematic)
```json
{
  "version": 2,
  "projects": {
    "booking-app": {
      "projectType": "application",
      "root": "apps/booking-app",
      "sourceRoot": "apps/booking-app/src",
      "prefix": "booking",
      "targets": {
        "build": {
          "executor": "@angular-devkit/build-angular:browser",
          "options": {
            "outputPath": "dist/apps/booking-app",
            "index": "apps/booking-app/src/index.html",
            "main": "apps/booking-app/src/main.ts",
            "polyfills": "apps/booking-app/src/polyfills.ts",
            "tsConfig": "apps/booking-app/tsconfig.app.json",
            "aot": true,
            "assets": [
              "apps/booking-app/src/favicon.ico",
              "apps/booking-app/src/assets"
            ],
            "styles": [
              "apps/booking-app/src/styles.css"
            ],
            "scripts": []
          },
          "configurations": {
            "production": {
              "fileReplacements": [
                {
                  "replace": "apps/booking-app/src/environments/environment.ts",
                  "with": "apps/booking-app/src/environments/environment.prod.ts"
                }
              ],
              "optimization": true,
              "outputHashing": "all",
              "sourceMap": false,
              "namedChunks": false,
              "extractLicenses": true,
              "vendorChunk": false,
              "buildOptimizer": true,
              "budgets": [
                {
                  "type": "initial",
                  "maximumWarning": "2mb",
                  "maximumError": "5mb"
                },
                {
                  "type": "anyComponentStyle",
                  "maximumWarning": "6kb",
                  "maximumError": "10kb"
                }
              ]
            }
          }
        },
        "serve": {
          "executor": "@angular-devkit/build-angular:dev-server",
          "options": {
            "browserTarget": "booking-app:build"
          },
          "configurations": {
            "production": {
              "browserTarget": "booking-app:build:production"
            }
          }
        },
        "extract-i18n": {
          "executor": "@angular-devkit/build-angular:extract-i18n",
          "options": {
            "browserTarget": "booking-app:build"
          }
        },
        "lint": {
          "executor": "@nrwl/linter:eslint",
          "options": {
            "lintFilePatterns": [
              "apps/booking-app/src/**/*.ts",
              "apps/booking-app/src/**/*.html"
            ]
          }
        },
        "test": {
          "executor": "@nrwl/jest:jest",
          "outputs": [
            "coverage/apps/booking-app"
          ],
          "options": {
            "jestConfig": "apps/booking-app/jest.config.js",
            "passWithNoTests": true
          }
        }
      }
    },
    "booking-app-e2e": {
      "root": "apps/booking-app-e2e",
      "sourceRoot": "apps/booking-app-e2e/src",
      "projectType": "application",
      "targets": {
        "e2e": {
          "executor": "@nrwl/cypress:cypress",
          "options": {
            "cypressConfig": "apps/booking-app-e2e/cypress.json",
            "tsConfig": "apps/booking-app-e2e/tsconfig.e2e.json",
            "devServerTarget": "booking-app:serve"
          },
          "configurations": {
            "production": {
              "devServerTarget": "booking-app:serve:production"
            }
          }
        },
        "lint": {
          "executor": "@nrwl/linter:eslint",
          "options": {
            "lintFilePatterns": [
              "apps/booking-app-e2e/**/*.{js,ts}"
            ]
          }
        }
      }
    },
    "booking-feature-flight-search": {
      "projectType": "library",
      "root": "libs/booking/feature-flight-search",
      "sourceRoot": "libs/booking/feature-flight-search/src",
      "prefix": "booking",
      "targets": {
        "build": {
          "executor": "@nrwl/angular:ng-packagr-lite",
          "options": {
            "tsConfig": "libs/booking/feature-flight-search/tsconfig.lib.json",
            "project": "libs/booking/feature-flight-search/ng-package.json"
          },
          "configurations": {
            "production": {
              "tsConfig": "libs/booking/feature-flight-search/tsconfig.lib.prod.json"
            }
          }
        },
        "lint": {
          "executor": "@nrwl/linter:eslint",
          "options": {
            "lintFilePatterns": [
              "libs/booking/feature-flight-search/src/**/*.ts",
              "libs/booking/feature-flight-search/src/**/*.html"
            ]
          }
        },
        "test": {
          "executor": "@nrwl/jest:jest",
          "outputs": [
            "coverage/libs/booking/feature-flight-search"
          ],
          "options": {
            "jestConfig": "libs/booking/feature-flight-search/jest.config.js",
            "passWithNoTests": true
          }
        }
      }
    }
  },
  "cli": {
    "defaultCollection": "@nrwl/angular",
    "packageManager": "yarn"
  },
  "generators": {
    "@nrwl/angular:application": {
      "unitTestRunner": "jest",
      "e2eTestRunner": "cypress",
      "linter": "eslint"
    },
    "@nrwl/angular:library": {
      "unitTestRunner": "jest",
      "linter": "eslint"
    },
    "@nrwl/angular:component": {}
  },
  "defaultProject": "booking-app"
}
```

`angular.json` version 1 (works with this schematic)
```json
{
  "version": 1,
  "projects": {
    "booking-app": {
      "projectType": "application",
      "root": "apps/booking-app",
      "sourceRoot": "apps/booking-app/src",
      "prefix": "booking",
      "architect": {
        "build": {
          "builder": "@angular-devkit/build-angular:browser",
          "options": {
            "outputPath": "dist/apps/booking-app",
            "index": "apps/booking-app/src/index.html",
            "main": "apps/booking-app/src/main.ts",
            "polyfills": "apps/booking-app/src/polyfills.ts",
            "tsConfig": "apps/booking-app/tsconfig.app.json",
            "aot": true,
            "assets": [
              "apps/booking-app/src/favicon.ico",
              "apps/booking-app/src/assets"
            ],
            "styles": ["apps/booking-app/src/styles.css"],
            "scripts": []
          },
          "configurations": {
            "production": {
              "fileReplacements": [
                {
                  "replace": "apps/booking-app/src/environments/environment.ts",
                  "with": "apps/booking-app/src/environments/environment.prod.ts"
                }
              ],
              "optimization": true,
              "outputHashing": "all",
              "sourceMap": false,
              "namedChunks": false,
              "extractLicenses": true,
              "vendorChunk": false,
              "buildOptimizer": true,
              "budgets": [
                {
                  "type": "initial",
                  "maximumWarning": "2mb",
                  "maximumError": "5mb"
                },
                {
                  "type": "anyComponentStyle",
                  "maximumWarning": "6kb",
                  "maximumError": "10kb"
                }
              ]
            }
          }
        },
        "serve": {
          "builder": "@angular-devkit/build-angular:dev-server",
          "options": {
            "browserTarget": "booking-app:build"
          },
          "configurations": {
            "production": {
              "browserTarget": "booking-app:build:production"
            }
          }
        },
        "extract-i18n": {
          "builder": "@angular-devkit/build-angular:extract-i18n",
          "options": {
            "browserTarget": "booking-app:build"
          }
        },
        "lint": {
          "builder": "@nrwl/linter:eslint",
          "options": {
            "lintFilePatterns": [
              "apps/booking-app/src/**/*.ts",
              "apps/booking-app/src/**/*.html"
            ]
          }
        },
        "test": {
          "builder": "@nrwl/jest:jest",
          "outputs": ["coverage/apps/booking-app"],
          "options": {
            "jestConfig": "apps/booking-app/jest.config.js",
            "passWithNoTests": true
          }
        }
      }
    },
    "booking-app-e2e": {
      "root": "apps/booking-app-e2e",
      "sourceRoot": "apps/booking-app-e2e/src",
      "projectType": "application",
      "architect": {
        "e2e": {
          "builder": "@nrwl/cypress:cypress",
          "options": {
            "cypressConfig": "apps/booking-app-e2e/cypress.json",
            "tsConfig": "apps/booking-app-e2e/tsconfig.e2e.json",
            "devServerTarget": "booking-app:serve"
          },
          "configurations": {
            "production": {
              "devServerTarget": "booking-app:serve:production"
            }
          }
        },
        "lint": {
          "builder": "@nrwl/linter:eslint",
          "options": {
            "lintFilePatterns": ["apps/booking-app-e2e/**/*.{js,ts}"]
          }
        }
      }
    },
    "booking-feature-flight-search": {
      "projectType": "library",
      "root": "libs/booking/feature-flight-search",
      "sourceRoot": "libs/booking/feature-flight-search/src",
      "prefix": "booking",
      "architect": {
        "build": {
          "builder": "@nrwl/angular:ng-packagr-lite",
          "options": {
            "tsConfig": "libs/booking/feature-flight-search/tsconfig.lib.json",
            "project": "libs/booking/feature-flight-search/ng-package.json"
          },
          "configurations": {
            "production": {
              "tsConfig": "libs/booking/feature-flight-search/tsconfig.lib.prod.json"
            }
          }
        },
        "lint": {
          "builder": "@nrwl/linter:eslint",
          "options": {
            "lintFilePatterns": [
              "libs/booking/feature-flight-search/src/**/*.ts",
              "libs/booking/feature-flight-search/src/**/*.html"
            ]
          }
        },
        "test": {
          "builder": "@nrwl/jest:jest",
          "outputs": ["coverage/libs/booking/feature-flight-search"],
          "options": {
            "jestConfig": "libs/booking/feature-flight-search/jest.config.js",
            "passWithNoTests": true
          }
        }
      }
    }
  },
  "cli": {
    "defaultCollection": "@nrwl/angular",
    "packageManager": "yarn"
  },
  "schematics": {
    "@nrwl/angular:application": {
      "unitTestRunner": "jest",
      "e2eTestRunner": "cypress",
      "linter": "eslint"
    },
    "@nrwl/angular:library": {
      "unitTestRunner": "jest",
      "linter": "eslint"
    },
    "@nrwl/angular:component": {}
  },
  "defaultProject": "booking-app"
}
```

**Comment by vsavkin** [comment]
From what I can see the issue is not related to v2 format. It's only related to the file name.

This is what throws the error:

```
function getWorkspacePath(host) {
    const possibleFiles = ['/angular.json', '/.angular.json'];
    const path = possibleFiles.filter(path => host.exists(path))[0];
    return path;
}
exports.getWorkspacePath = getWorkspacePath;
function getWorkspace(host) {
    const path = getWorkspacePath(host);
    const configBuffer = host.read(path);
    if (configBuffer === null) {
        throw new schematics_1.SchematicsException(`Could not find (${path})`);
    }
    const content = configBuffer.toString();
    return core_1.parseJson(content, core_1.JsonParseMode.Loose);
}
```

It's unable to find the configuration cause it's only looking for `angular.json`. The reason why it's only looking for anguar.json is that the package is not updated. It depends on Angular 7 and schematics v7, which didn't look for `workspace.json`.

We have a custom host what converts the configuration back to v1. For instance, the angular library schematic reads workspace.json, and, if it's v2, it will be converted to the v1 format so the schematic works.

**Comment by LayZeeDK** [comment]
Thanks for loooking into this with your expert knowledge, @vsavkin.

I will open an issue with `ng2-charts-schematics`.

Is the code listed from angular@7, angular-devkit@7, or `ng2-charts-schematics`?

**Comment by vsavkin** [comment]
@LayZeeDK I think this might fix it: https://github.com/nrwl/nx/pull/4369 but I need to do a lot more testing. The community and the angular team themselves don't use the config stuff consistently, so I'm always scared to break some edge case where some migration for some weird reason don't do the right thing etc.

Thanks for reporting the issue!

**Comment by LayZeeDK** [comment]
Just had a similar issue with the `@ngrx/store:ng-add` schematic
```
PS C:\projects\github\nx-worker\nxworker> nx generate @ngrx/store:ng-add
SchematicsException [Error]: Could not find (undefined)
    at Object.getWorkspace (C:\projects\github\nx-worker\nxworker\node_modules\.pnpm\@ngrx\store@10.1.2_@angular+core@11.0.5+rxjs@6.5.5\node_modules\@ngrx\store\schematics-core\utility\config.js:15:15)
    at getProject (C:\projects\github\nx-worker\nxworker\node_modules\.pnpm\@ngrx\store@10.1.2_@angular+core@11.0.5+rxjs@6.5.5\node_modules\@ngrx\store\schematics-core\utility\project.js:6:30)
    at Object.getProjectPath (C:\projects\github\nx-worker\nxworker\node_modules\.pnpm\@ngrx\store@10.1.2_@angular+core@11.0.5+rxjs@6.5.5\node_modules\@ngrx\store\schematics-core\utility\project.js:17:19)
    at C:\projects\github\nx-worker\nxworker\node_modules\.pnpm\@ngrx\store@10.1.2_@angular+core@11.0.5+rxjs@6.5.5\node_modules\@ngrx\store\schematics\ng-add\index.js:108:42
    at MergeMapSubscriber.project (C:\projects\github\nx-worker\nxworker\node_modules\.pnpm\@angular-devkit\schematics@11.0.5\node_modules\@angular-devkit\schematics\src\rules\call.js:75:24)
    at MergeMapSubscriber._tryNext (C:\projects\github\nx-worker\nxworker\node_modules\.pnpm\rxjs@6.6.3\node_modules\rxjs\internal\operators\mergeMap.js:67:27)
    at MergeMapSubscriber._next (C:\projects\github\nx-worker\nxworker\node_modules\.pnpm\rxjs@6.6.3\node_modules\rxjs\internal\operators\mergeMap.js:57:18)
    at MergeMapSubscriber.Subscriber.next (C:\projects\github\nx-worker\nxworker\node_modules\.pnpm\rxjs@6.6.3\node_modules\rxjs\internal\Subscriber.js:66:18)
    at Observable._subscribe (C:\projects\github\nx-worker\nxworker\node_modules\.pnpm\rxjs@6.6.3\node_modules\rxjs\internal\util\subscribeToArray.js:5:20)
    at Observable._trySubscribe (C:\projects\github\nx-worker\nxworker\node_modules\.pnpm\rxjs@6.6.3\node_modules\rxjs\internal\Observable.js:44:25)
Could not find (undefined)
```

`@ngrx/store/schematics-core/utility/config.js` from `@ngrx/store@10.1.2`

```ts
"use strict";
exports.__esModule = true;
exports.getWorkspace = exports.getWorkspacePath = void 0;
var schematics_1 = require("@angular-devkit/schematics");
function getWorkspacePath(host) {
    var possibleFiles = ['/angular.json', '/.angular.json']; // 👈
    var path = possibleFiles.filter(function (path) { return host.exists(path); })[0];
    return path;
}
exports.getWorkspacePath = getWorkspacePath;
function getWorkspace(host) {
    var path = getWorkspacePath(host); // 👈
    var configBuffer = host.read(path);
    if (configBuffer === null) {
        throw new schematics_1.SchematicsException("Could not find (" + path + ")"); // 👈
    }
    var config = configBuffer.toString();
    return JSON.parse(config);
}
exports.getWorkspace = getWorkspace;
//# sourceMappingURL=config.js.map
```👈

**Comment by vsavkin** [comment]
This has been fixed in 11.1.1 (at least in the packages I've tried). Thanks again for reporting this!
